### ISSUE TITLE
Move map and bindings entries

### ISSUE THREAD

**Issue by mainej** [clojure-lsp/clojure-lsp#684]
When I'm editing a map I often want to move one key/value pair before the prior pair (or after the next pair). Similarly, in a let binding, I often want to move a binding form and its value before the prior form (or after the subsequent form). This is a rote transformation, but it often takes many steps. This is especially true if the first or last pair is changing, because line-wise changes in the editor tend to break the structure of the map or binding, because they move the opening or closing brackets.

For example, if I'm on the `:b 2` line in this form:

```clojure
            {:a 1
             :b 2
             :c 3}
```
I'd like to be able to move it up, like so:
```clojure
            {:b 2
             :a 1
             :c 3}
```
Or down, like so:
```clojure
            {:a 1
             :c 3
             :b 2}
```

Ideally, comments would stay "affiliated" to their respective pairs. For example, consider the comments related to `:b` below, both on the line before it and on the same line, at the end.
```clojure
            {:a 1 ;; comment after a
             ;; comment before b
             :b 2 ;; comment after b
             ;; multiline
             ;; comment for c
             :c 3}
```
This would become (moving up):
```clojure
            {;; comment before b
             :b 2 ;; comment after b
             :a 1 ;; comment after a
             ;; multiline
             ;; comment for c
             :c 3}
```
Or (moving down):
```clojure
            {:a 1 ;; comment after a
             ;; multiline
             ;; comment for c
             :c 3
             ;; comment before b
             :b 2 ;; comment after b
             }
```

> Note an edge case demonstrated in the last example: because of the comment `;; comment after b`, the closing bracket of the map needs to be put on its own line, to avoid being commented out.

Would you be willing to add this refactoring to clojure-lsp? Or is there an established way to extend the code actions that clojure-lsp exposes?

I have a prototype that does the described refactoring. It preserves affiliated comments, as described above. It uses `clj-rewrite` and operates on strings, but doesn't have a full editor integration which is why I'm asking here. I'm not sure it's very idiomatic use of `clj-rewrite`, but it at least proves it's possible. Anyway, I'd be happy to share that code.

**Comment by ericdallo** [comment]
@mainej It's been on my mind to add this feature for some time but I never had the time, the issue with that is the comments manipulation, if you have a code that provides this feature and want to share, I'd make things a lot easier :)

**Comment by mainej** [comment]
@ericdallo I'm happy to share... see below. 

If you'd prefer I can make a PR, though I might need some guidance. That is, this code is functional on its own, but I'm not familiar enough with clojure-lsp to know how to convert it into an action.

I've included a few examples in a rich comment block, but it needs proper unit tests. There are several edge cases that the code handles but which don't have an example. That'd be another reason to make a PR—so I could ensure those keep working.

There are a few places marked TODO which still need behavior defined.

```clojure
(ns scratch
  (:require [rewrite-clj.node :as n]
            [rewrite-clj.zip :as z]
            [clojure.string :as string]))

(defn- expand-comment-newlines
  "Comment nodes often include their trailing newline. This splits such nodes
  into two, to allow more careful handling of newlines."
  [children]
  (mapcat (fn [{:keys [s prefix] :as n}]
            (if (and (n/comment? n)
                     (string/ends-with? s "\n"))
              [(n/comment-node prefix (subs s 0 (dec (count s))))
               (n/newline-node "\n")]
              [n]))
          children))

(defn- whitespace-comment-or-newline? [node]
  (contains? #{:whitespace :comment :newline} (n/tag node)))

(defn- whitespace-or-comment? [node]
  (contains? #{:whitespace :comment} (n/tag node)))

(defn- parse-entry-pairs
  "Parse the children of a map or binding `parent-zloc` into entry pairs
  (key/value pairs or binding/value pairs respectively).

  An entry pair is vector of nodes: the key and value themselves, with their
  surrounding whitespace, comments and newlines.

  Logically, a pair's format is:
  (whitespace | comment | newline)*
  key
  (whitespace | comment | newline)*
  value
  (whitespace | comment)*

  NOTE: A pair does NOT include any trailing newlines... they are all allocated
  to the next pair.

  Returns two values: a vector of the entry pairs, and if there are lines
  after the last value, a vector of those additional nodes."
  [parent-zloc]
  (let [[elements extra-lines]
        (loop [children (->> parent-zloc
                             z/node
                             n/children
                             expand-comment-newlines)
               state    :before-key
               result   []]
          (cond
            (not (seq children))
            ;; everything processed
            [result []]

            (every? whitespace-comment-or-newline? children)
            ;; everything processed but extra lines
            [result children]

            (= :before-key state)
            ;; gather comments before key, and key itself
            (let [[prefix [k & rst]] (split-with whitespace-comment-or-newline?
                                                 children)]
              (recur rst
                     :before-val
                     (conj result (concat prefix (when k [k])))))

            (= :before-val state)
            ;; gather comments before val, val itself, and optional comment
            ;; after val on same line
            (let [[prefix [v & rst]] (split-with whitespace-comment-or-newline?
                                                 children)
                  [postfix rst]      (split-with whitespace-or-comment?
                                                 rst)]
              (recur rst
                     :before-key
                     (conj result (concat prefix (when v [v]) postfix))))))]
    [(->> elements
          (partition-all 2)
          (mapv (fn [[k v]] (concat k v))))
     extra-lines]))

(defn- move-entry [zloc dir]
  ;; TODO: abort if the zloc is on whitespace or a comment. Weird swaps happen.
  ;; TODO: abort if the parent contains an odd number of elements, we aren't in
  ;; a valid map or binding, and weird swaps will happen.
  (let [parent-zloc (z/up zloc)]
    (if-not (and parent-zloc
                 ;; We want to support 'things with paired entries': maps as well
                 ;; as let bindings and other binding forms. Rather than try to
                 ;; detect whether something is a binding form, we just treat all
                 ;; vectors as binding forms.
                 (contains? #{:map :vector}
                            (n/tag (z/node parent-zloc))))
      zloc
      (let [[entry-pairs extra-lines] (parse-entry-pairs parent-zloc)

            ;; Figure out which pairs to swap
            node-position     (->> zloc
                                   (iterate z/left)
                                   (take-while (complement z/leftmost?))
                                   count)
            new-node-position (dir (dir node-position))
            ;; position of the entry pair we're moving, whether the original
            ;; node was a key or value.
            position          (-> node-position (/ 2) int)
            new-position      (dir position)]
        (if-not (contains? entry-pairs new-position)
          ;; trying to move first entry up or last down, which should be a no-op
          zloc
          (let [;; Swap entries.
                entry-pairs (assoc entry-pairs
                                   position (get entry-pairs new-position)
                                   new-position (get entry-pairs position))
                ;; Most entries can keep the whitespace that precedes them,
                ;; whether moving up or down. This assumption fails when
                ;; swapping the first and second entries. The entry that is
                ;; initially first typically doesn't have any preceding
                ;; whitespace, but it's moving to a location where it needs
                ;; some. And conversely, the entry that is initially second
                ;; typically has preceding whitespace but is moving somewhere
                ;; where it shouldn't. We fix this by swapping preceding
                ;; whitespace between the two.
                ;; TODO: seems to work whether or not we're swapping the first
                ;; and second entries. But is that accounting for all whitespace
                ;; edge cases? Maybe better to do this only if (= #{0 1}
                ;; (hash-set position new-position)).
                entry-pairs (-> entry-pairs
                                (update 0 #(drop-while n/whitespace? %))
                                (update 1 #(concat (take-while n/whitespace? (get entry-pairs 0))
                                                   %)))

                ;; Put revised entries back into parent.
                parent-zloc (->> (concat (flatten entry-pairs)
                                         extra-lines)
                                 (n/replace-children (z/node parent-zloc))
                                 (z/replace parent-zloc))
                ;; If after reordering, the last component has become a comment,
                ;; we need to add a newline or else the closing bracket will be
                ;; commented out.
                parent-zloc (if (-> parent-zloc z/down z/rightmost* z/node n/comment?)
                              (let [[_row col] (-> parent-zloc z/down z/rightmost z/left z/position)]
                                (-> parent-zloc
                                    z/down
                                    z/rightmost*
                                    (z/insert-space-right (dec col))
                                    z/insert-newline-right
                                    z/up))
                              parent-zloc)]
            ;; Move zipper back to original node, so repeated invocations keep
            ;; moving the same entry.
            (->> parent-zloc z/down (iterate z/right) (drop new-node-position) first)))))))

(defn move-entry-down [loc]
  (move-entry loc inc))

(defn move-entry-up [loc]
  (move-entry loc dec))

(comment
  (def map-form
    "            {:a 1 ;; comment after a
             ;; comment before b
             :b 2 ;; comment after b
             ;; multiline
             ;; comment for c
             :c 3}")

  (def root-loc (z/of-string map-form {:track-position? true}))

  (def aloc
    (z/down root-loc))

  (def bloc
    (-> aloc z/right z/right))

  (def cloc
    (-> bloc z/right z/right))

  (defn inspect [z]
    (println (z/root-string z)))

  (inspect (move-entry-down aloc))
  (inspect (move-entry-down (z/right aloc))) ;; works when cursor is on val too
  (inspect (move-entry-down bloc))
  (inspect (move-entry-down cloc)) ;; no change when moving last entry down

  (inspect (move-entry-up aloc)) ;; no change when moving first entry up
  (inspect (move-entry-up bloc))
  (inspect (move-entry-up cloc))

  ;; repeated invocations move further
  (-> aloc
      move-entry-down
      move-entry-down
      inspect))
```

**Comment by ericdallo** [comment]
@mainej thanks, I can make the integration with action and leave to you to add all missing tests then, WDYT?
I'll take a look at the code and let you know if I find any issues

**Comment by mainej** [comment]
@ericdallo sounds great. Thanks!

**Comment by snoe** [comment]
Ideally this should be content aware and transpose entries while in a map and elements in other collections/forms. This would be more similar to cursive's handling.

**Comment by ericdallo** [comment]
Not sure I got it, what would be different from the issue description @snoe ?

**Comment by mainej** [comment]
> elements in other collections

@snoe Are you saying that this should move individual nodes in a set, list, or regular vector, but pairs of nodes in a map or binding vector? 

That could be cool, but a bit tricky. Any suggestions for how to tell whether a vector is actually a binding? That is, I'd want this:
```clojure
(let [a 1
      b 2] ; <- move pairs in this vector
  [a
   b] ; <- move individuals in this vector
  )
```

In clojure.core certain symbols imply that you're working with a binding. I'm aware of `let`, `binding`, `with-redefs` and `with-local-vars` but there may be others. I could special case those, but what about library functions that establish bindings such as `reagent.core/with-let`? Also, app developers might write their own functions that establish bindings. Is there a way to register these with clojure-lsp so that they could also be handled as binding vectors?

**Comment by ericdallo** [comment]
@mainej maybe just leave vector as it is, sets and lists don't have pairs, sounds like a start

**Comment by mainej** [comment]
I'm getting a little confused. Can we call vectors that establish bindings "binding-vectors" and regular vectors "ordinary-vectors"?

@ericdallo do you mean handle maps, but don't try to handle vectors at all? I'd really like to re-order bindings too.

Or do you mean always treat ordinary-vectors like binding-vectors, moving their elements as pairs regardless? That could be a good first pass, and we could modify the behavior later.

**Comment by ericdallo** [comment]
Sorry for the miss-understand:

> Or do you mean always treat ordinary-vectors like binding-vectors, moving their elements as pairs regardless? That could be a good first pass, and we could modify the behavior later.

Yes, only if the children count of the vector is even we allow that as well.

I'm starting implementing the code action support and let you know soon

**Comment by mainej** [comment]
> only if the children count of the vector is even

@ericdallo OK, sounds good. Is your plan to add the code action support on a branch and hand it over to me to write tests?

Also, what I gave you works, but I think it would benefit from a bit more hammock time. I have a sense I can simplify the comment and newline handling. My hope is to make it work even when the cursor is on a comment, which only works in certain scenarios right now. I'll write tests around that and see if I can get them passing.

**Comment by ericdallo** [comment]
Sure, I'll do it in a branch so we can work together on that

**Comment by mainej** [comment]
Perfect. Thanks!

**Comment by ericdallo** [comment]
@mainej I created https://github.com/clojure-lsp/clojure-lsp/pull/685 which is already working 🎉 
Still need to add code actions tests, docs, changelog but it's enough for you to take a look and change what you think it makes sense or add more tests/cases, thank you for the huge help on this

![move-entry](https://user-images.githubusercontent.com/7820865/148492088-13d6dcc5-e7a4-4e16-ba49-9179dff78081.gif)

**Comment by mainej** [comment]
@ericdallo that's so cool! I'll check out the PR branch tomorrow and start making adjustments. 

Do we have control over where the cursor ends up? Or does it stay in the same row/column as it started in?

**Comment by ericdallo** [comment]
@mainej yes, it stay on the same position, unless we send another notification to move the cursor to another URI or range. We could send to move the cursor to the new entry position, I'll take a look and see if this would be possible/hard

**Comment by mainej** [comment]
Implemented in #685
